tools/trainerproc/main.c

```diff
@@ -0,0 +1,1965 @@
+        {
+            if (trainer->pic_line)
+                any_error = !set_show_parse_error(p, key.location, "duplicate 'Pic'");
+            trainer->pic_line = value.location.line;
+            trainer->pic = token_string(&value);
+        }
+        else if (is_literal_token(&key, "Name"))
+        {
+            if (trainer->name_line)
+                any_error = !set_show_parse_error(p, key.location, "duplicate 'Name'");
+            trainer->name_line = value.location.line;
+            trainer->name = token_string(&value);
+        }
+        else if (is_literal_token(&key, "Double Battle"))
+        {
+            if (trainer->double_battle_line)
+                any_error = !set_show_parse_error(p, key.location, "duplicate 'Double Battle'");
+            trainer->double_battle_line = value.location.line;
+            if (!token_bool(p, &value, &trainer->double_battle))
+                any_error = !show_parse_error(p);
+        }
+        else if (is_literal_token(&key, "Mugshot"))
+        {
+            if (trainer->mugshot_line)
+                any_error = !set_show_parse_error(p, key.location, "duplicate 'Mugshot'");
+            trainer->mugshot_line = value.location.line;
+            trainer->mugshot = token_string(&value);
+        }
+        else if (is_literal_token(&key, "Starting Status"))
+        {
+            if (trainer->starting_status_line)
+                any_error = !set_show_parse_error(p, key.location, "duplicate 'Starting Status'");
+            trainer->starting_status_line = value.location.line;
+            trainer->starting_status = token_string(&value);
+        }
+        else
+        {
+            any_error = !set_show_parse_error(p, key.location, "expected one of 'Name', 'Class', 'Pic', 'Gender', 'Music', 'Items', 'Double Battle', or 'AI'");
+        }
+    }
+    if (!trainer->pic_line)
+        any_error = !set_show_parse_error(p, p->location, "expected 'Pic' before Pokemon");
+    if (!trainer->name_line)
+        any_error = !set_show_parse_error(p, p->location, "expected 'Name' before Pokemon");
+    if (!match_empty_line(p))
+    {
+        set_show_parse_error(p, p->location, "expected empty line");
+
+        // Continue if the line is the start of a new Pokemon.
+        struct Parser p_ = *p;
+        if (!parse_pokemon_header(&p_, &nickname, &species, &gender, &item))
+            return false;
+    }
+
+    for (int i = 0; i < PARTY_SIZE; i++)
+    {
+        struct Pokemon *pokemon = &trainer->pokemon[i];
+
+        // Parse first line.
+        while (match_empty_line(p)) {}
+        if (!parse_pokemon_header(p, &nickname, &species, &gender, &item))
+        {
+            if (i > 0 || is_literal_string(trainer->id, "TRAINER_NONE"))
+                break;
+            if (!p->error)
+                set_parse_error(p, p->location, "expected nickname or species");
+            show_parse_error(p);
+
+            // TODO: Continue if a Pokemon starts before the next section.
```

Comment:
Please add this TODO to the description.

Reply:
I'm not sure how to phrase this one :thinking:

Reply:
I actually don't know what you mean by this either 😂

Reply:
I think it's something like "if a particular kind of parse error is encountered here, we should try to recover by starting parsing a new Pokémon". Just a minor detail for giving more focused error messages, a bit like making a C compiler assume that if it sees a new function start with no indentation before a `}` that a `}` is missing rather than spew a billion errors.